File: backend/backend/Models/NTuple.cs
using System.Linq;
using System;
using System.Collections.Generic;
using backend.DTO;

namespace backend.Models
{
    public class NTuple
    {
        public List<IInput> List { get; set; } = new ();

        public bool IntersectsWith(NTuple other)
        {
            if (List.Count != other.List.Count)
                return false;

            return List.Zip(other.List)
                       .All(x => x.First.IntersectsWith(x.Second));
        }

        public override bool Equals(object? obj) => Equals(obj as NTuple);

        private bool Equals(NTuple? that)
        {
            if (that == null || GetType() != that.GetType())
                return false;

            if(ReferenceEquals(that, this))
                return true;
            
            return this.List.SequenceEqual(that.List);
        }

        public override int GetHashCode() =>
            List.Aggregate(0, (item, hashCode) => HashCode.Combine(item.GetHashCode(), hashCode));

        public override string ToString()
        {
            var elements = String.Join(",", List);
            return $"({elements})";
        }
    }
}

File: frontend/src/logic/models/ntuple.ts
import type { IInput } from "./dtos";
import { zip } from 'fp-ts/Array';

export class NTuple {
    list: IInput[];

    constructor(list: IInput[]) {
        this.list = list;
    }

    intersectsWith(other: NTuple): boolean {
        if(this.list.length !== other.list.length)
            return false;

        return zip(this.list, other.list)
                .every(([x,y]) => x.intersectsWith(y));
    }
}
